Initial PDB / PerfMap support in Crossgen2 + System.Private.CoreLib switchover to use Crossgen2

trylek opened the pull request:
(*) Move PDB writer support out of R2RDump to a new assembly ILCompiler.Diagnostics;

(*) Add initial PerfMap writer support parallel to PDB emitter;

(*) Add Crossgen2 support for emitting PDB / PerfMap using the ILCompiler.Diagnostics assembly;

(*) Add support for PDB generation to R2RTest;

(*) Add support for optional PDB generation to the src\tests\build.cmd script (PDB is only supported on Windows).

Thanks

Tomas

trylek committed:
Move R2RDump PDB writer code to a separate assembly

I have moved the PDB writer code to a new assembly
ILCompiler.Diagnostics so that it can be reused by Crossgen2.
I have also added an initial trivial implementation of the
PerfMap writer.

Thanks

Tomas

trylek committed:
Initial changes for PDB generation support in Crossgen2

trylek committed:
Crossgen2 implementation of PDB / Perfmap generator

trylek committed:
R2RTest / test build support for PDB production

trylek commented:
/cc @dotnet/crossgen-contrib

mangod9 commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/ObjectWriter/MapFileBuilder.cs:
```diff
@@ -411,5 +446,55 @@ private void WriteTitle(StreamWriter writer, string title)
+        }
     }
 }
+;
```
is this a leftover ';'?

trylek replied:
Weird, no idea how it got there, thanks for spotting, removed in updated commit.

trylek committed:
Typo in perfmap option, superfluous semicolon spotted by Manish

nattress commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/ObjectWriter/MapFileBuilder.cs:
```diff
@@ -205,6 +226,20 @@ public void SaveCsv(string nodeStatsCsvFileName, string mapCsvFileName)
             }
         }
 
+        public void SavePdb(string pdbPath, string dllFileName)
```
It feels weird having `MapFileBuilder` be the thing that writes PDBs and Perf maps since its purpose is to write a text summary file at the end of compilation. The map file builder's data structures at this point are a nice model of node to file location which makes it handy for these purposes. Maybe we could split the concepts into a class containing mapping details (shareable by map, perfmap, pdb generation)?

trylek replied:
Fixed in 9th commit - I have split the logic into an "output file info collector" (OutputInfoBuilder) and the actual emitters using it - MapFileBuilder and SymbolFileBuilder, thanks for the suggestion.

trylek committed:
Better align parameters for PDB / PerfMap generation

mangod9 commented:
Appears there are still some lingering issues on x86?

trylek committed:
Add previously missing project properties to ILCompiler.Diagnostics.csproj

trylek committed:
Addressing Simon's PR feedback, msbuild project fixes for x86

Based on Simon's suggestion I moved the PDB / PerfMap-specific code
to a new file SymbolFileBuilder as I concur that its squatting in
the MapFileBuilder was somewhat hacky. Both classes use a new helper
class ObjectInfoBuilder that collects the information common to
both - I was reluctant to make it a base class of the *FileBuilder's
as we'd have to collect all the elements twice.

Thanks

Tomas

trylek committed:
Rename ObjectItem/Node/InfoBuilder to Output* to fix ambiguity

trylek commented:
I have addressed initial Manish's and Simon's PR feedback and the change finally passes PR testing. I have launched an outerloop run and I believe the change is ready to be merged in. Please take another look when you have a chance and let me know what additional comments need addressing before this goes in.

nattress reviewed (approved):
This looks awesome to me 😎 Thanks Tomas!
